Improvements to the Russian computer braille table

Futyn-Maker opened the pull request:
Hello!

I've added new characters for Bashkir and Uzbek languages. Now it is possible to read texts on these languages by this table. @tc-dl please check the correctness.

Andrey

Futyn-Maker commented:
@bertfrees I think You can merge it. I've taken the definitions from corresponding tables, so, it should be correct.

tc-dl commented:
I'll check back later. will these Bashkir braille tables be published?

bertfrees commented:
@Futyn-Maker Can you explain how this relates to the other PR?

Futyn-Maker commented:
> Can you explain how this relates to the other PR (the Bashkir one)?

We figure out that it doesn't make sense to add the Bashkir computer Braille table, because it's impossible to provide the back translation of the special Bashkir symbols. I've decided to add this symbols into Russian computer Braille table to provide reading (forward translation) of these. As a bonus, I've added special symbols for Uzbek language.

bertfrees commented:
So can I take the tests from the other PR and apply them here?

tc-dl commented:
yes, you don't need to add Bashkir computer braille. Will the literature table be published?

bertfrees commented:
I don't know. Let's hope so.

Futyn-Maker commented:
@bertfrees I've added the missing symbols in the table and updated the test-file adding test for Bashkir symbols. I think You need to delete the Bashkir computer table and the test for it.

egli commented:
Move tests from #1005 to this PR.

Futyn-Maker commented:
@egli I think all You need is to remove tests for computer Braille table from https://github.com/liblouis/liblouis/pull/1005. I've already added test for Bashkir symbols to ru.yaml.

JuniGobi commented:
Hello!

What should be done next? Who can fix the conflict in the files? Please help us.

bertfrees commented:
@JuniGobi I'm taking care of it. If I have more questions I will let you know.

bertfrees commented:
@Futyn-Maker I'd rather not just remove those tests, I'd like to keep them and run them against your improved Russian table. I just tried this and I'm getting the following failures. Can you and @tc-dl clarify? Can you tell me for each of these whether (a) the failure is expected because of limitations of the table, (b) there is a mistake in the table, (c) the test is wrong, or (d) the Russian and Bashkir braille codes differ in these areas.

The first issue is a back-translation issue. The forward translation is correct:

Input:    '⡅⠣⠬⠣⠝⠀⠅⠣⠝⠙⠜⠗⠀⠧⠮⠥⠮⠝⠁⠠⠀⠫⠏⠗⠁⠩⠞⠁⠗⠀⠧⠁⠗⠻⠁⠫⠠⠀⠅⠜⠯⠑⠋⠞⠜⠗⠀⠧⠁⠻⠮⠱⠩⠁⠀⠞⠁⠗⠞⠮⠇⠁⠄'
Expected: 'Көҙөн көндәр һыуына, япраҡтар һарғая, кәйефтәр һағышҡа тартыла.'
Received: 'К(`(н к(нд)р ўыуына, япра%тар ўар~ая, к)йефт)р ўа~ыш%а тартыла.'

The two next examples are not forward translated correctly:

Input:    '(~)'
Expected: '⢔⢶⡢'
Received: '⠣⠻⠜'

Input:    'example@domain.com'
Expected: '⢑⢭⢁⢍⢏⢇⢑⣔⢙⢕⢍⢁⢊⢝⠄⢉⢕⢍'
Received: '⢑⢭⢁⢍⢏⢇⢑⡜⢙⢕⢍⢁⢊⢝⠄⢉⢕⢍'

Thanks.

Futyn-Maker commented:
@bertfrees Hello!

The problem is that @JuniGobi seems to base the Bashkir test on uk.yaml. It doesn't contains any unique tests excepted test № 1 in Your previous comment, I've included this test in ru.yaml. So we have following:
1. It was figured out that we can't provide the back translation of Bashkir symbols, so the first test should be only with forward flag;
2. The remaining failur tests are because of changing punctuations in Ukrainian computer Braille table. As I've written above, this Bashkir test was totally copied from uk.yaml without any changes, excepted adding test for Bashkir symbols as the first test.

According to this facts, I suggest following:
* remove all tests connected with Bashkir computer Braille table from ba.yaml as it doesn't contains any unique tests anymore;
* Remove the Bashkir computer Braille table at all (we've discussed it above in this issue);
* Add the Bashkir literary table with tests for it;
* Merge this PR. :-)

Thanks and sorry for confusions, I just want to do it better.

Andrey

bertfrees commented:
> It was figured out that we can't provide the back translation of Bashkir symbols, so the first test should be only with forward flag;

OK. This is clearly not optimal though, right? How useful is a braille code if there are letter in the alphabet that don't have a dedicated braille pattern?

> remove all tests connected with Bashkir computer Braille table from ba.yaml as it doesn't contains any unique tests anymore;

I don't think it's a real problem if the same tests exists in multiple braille-specs files, if those files are for different languages. If needed we can clean this up later. I'd rather keep a few too many tests than throw away relevant stuff.

It is a however a problem if test data was committed that is not correct. That's why I want to double check with @tc-dl and @JuniGobi . Please tell me, was this a mistake or intended?

Futyn-Maker commented:
> OK, will do. This is clearly not optimal though, right? How useful is a braille code if there are letter in the alphabet that don't have a dedicated braille pattern?

We have what we have... The problem is because of conflicts with punctuations. There aren't any document which discribes the Bashkir Braille code. Impossibility of providing the back translation is cause I suggest removing the Bashkir computer Braille table and adding this symbols in the Russian table. Bashkir-language-speaking people could read their texts by using the Russian table (we speak about computer Braille code).

> I don't think it's a real problem if the same tests exists in multiple braille-specs files, if those files are for different languages. If needed we can clean this up later. I'd rather keep a few too many tests than throw away relevant stuff.

Of course, but if we aren't going to include ba-comp.utb, what does these tests make sense? You seem couldn't connect it with any table.

But if You'd prefer not to remove these ones, it's necessary to do two things:
* Move the first test into forward section;
* Remove tests which I've created specially for Ukrainian computer Braille table. TThey aren't relevant for Bashkir tables...

bertfrees commented:
> Of course, but if we aren't going to include ba-comp.utb, what does these tests make sense? You seem couldn't connect it with any table.

I connect the test with the Russian computer table for now.

> Remove tests which I've created specially for Ukrainian computer Braille table. They aren't relevant for Bashkir tables...

No, I don't think so either, but allow me to double check this with the people that committed the test file. And if it was indeed a mistake, like we both suspect, the test should be corrected rather than removed.

Futyn-Maker commented:
> No, I don't think so either, but allow me to double check this with the people that committed the test file. And if it was indeed a mistake, like we both suspect, the test should be corrected rather than removed.

OK. But I think these tests below are excessive at all:
```Input: '(~)'
Expected: '⢔⢶⡢'
Received: '⠣⠻⠜'
Input: 'example@domain.com'
Expected: '⢑⢭⢁⢍⢏⢇⢑⣔⢙⢕⢍⢁⢊⢝⠄⢉⢕⢍'
Received: '⢑⢭⢁⢍⢏⢇⢑⡜⢙⢕⢍⢁⢊⢝⠄⢉⢕⢍'```
They were created specially to fix changing of punctuations in the Ukrainian Braille table. There're no any changes in Bashkir code.

bertfrees commented:
It doesn't matter. It improves the test coverage. It would also be better if the Russian test file would have tests for these characters.

Futyn-Maker commented:
Well, it's possible to correct the Bashkir test by copying tests from ru.yaml which I've made now.

tc-dl commented:
yes, all files were made from Ukrainian table. the test file was added last. To be honest, I did not quite understand it, but I also tried to do it by analogy. I created braille tables after reading the documentation through a translator because I practically do not speak English, so I think that maybe I missed something, or I misunderstood. what is missing in the test file?

bertfrees commented:
@Futyn-Maker Great, thanks.

@tc-dl OK thanks, that explains it. You can just tell me in all honesty when you don't fully understand something. I'm just gonna go with whatever Andrey tells me to do now. If there are issues with the table you can just talk to him, then there is also no language barrier.

Futyn-Maker commented:
Yes, I know Russian and can answer the questions here or by [mailto:andrewia2002@yandex.ru](E-Mail). @bertfrees I think correcting the Bashkir test is the last thing we need, right?

Andrey

bertfrees commented:
Yes I'm finalizing things now.

bertfrees commented:
Replaced by https://github.com/liblouis/liblouis/pull/1027.
